examples/fabric-chaincode-example-external/src/main/java/org/hyperledger/fabric/example/Application.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2019 IBM All Rights Reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.hyperledger.fabric.example;
+
+import org.hyperledger.fabric.shim.ChaincodeServer;
+import org.hyperledger.fabric.shim.ChaincodeServerImpl;
+
+import java.io.IOException;
+
+public class Application {
```

Comment:
So is this the main entry point? This needs to be started before connecting to the peer.
Is there a reference to the main documentation on how the peer should be configured if required.

Reply:
> @andreevym that would be helpful - but let me review the changes first.
> 
> Are there any updates planned to the main documentation? I can see there's an example here which is good.

I want to add documentation for write external chaincode with fabric-chaincode-java, but in another PR

Reply:
> So is this the main entry point? This needs to be started before connecting to the peer.
> Is there a reference to the main documentation on how the peer should be configured if required.

yes, the Application main class in external java chaincode

we need to run application with following env variable (with our value)

CORE_CHAINCODE_ID_NAME=externalcc:06d1d324e858751d6eb4211885e9fd9ff74b62cb4ffda2242277fac95d467033;
     * CORE_PEER_ADDRESS=127.0.0.1:7051;
     * CORE_PEER_TLS_ENABLED=false;
     * CORE_PEER_TLS_ROOTCERT_FILE=src/test/resources/ca.crt;
     * CORE_TLS_CLIENT_KEY_PATH=src/test/resources/client.key.enc;
     * CORE_TLS_CLIENT_CERT_PATH=src/test/resources/client.crt.enc;
     * PORT_CHAINCODE_SERVER=9999;

Reply:
@andreevym we should move away from ENV vars into using the ChaincodeServer class to encapsulate (as in https://github.com/hyperledger/fabric-chaincode-go/blob/master/shim/chaincodeserver.go#L27). This will help in a few ways
* let user determine her input model (we don't force user to use ENVs but they can in _their_ chaincode program)
* further isolate the old from the new (we should try to reuse as much of the "processing" path. The more we isolate at the "connection" point, more it'll be easier to reuse). For example https://github.com/hyperledger/fabric-chaincode-go/blob/master/shim/internal/config.go#L101 was created as a common method to so we dont have to rely on older model's config handling https://github.com/hyperledger/fabric-chaincode-go/blob/master/shim/internal/config.go#L28  

Please note that the above code samples are just to illustrate ideas and may not translate directly to the code.

Reply:
With the context of the Java repo, I don't believe creating a class to hold these properties is required. True the current implementation uses both environment variables and system properties, but we should look to be more rigourous here.

Reply:
`but we should look to be more rigourous here.`

Perhaps we are saying the same thing but just for clarity I interpret that as follows : the key diff between the two models is how the runtime lifecycle is handled. It is transparent to the user in the prev model (fabric is responsible) but now her full responsibility. We should avoid imposing constraints such as needing environment variables. Indeed this is the whole point behind exposing a type to program. By the way type was not strictly required in the GO repo as well but the above considerations drove us towards it (initiated by @mastersingh24 to give due credit :-) ).

Reply:
@muralisrini  

Agreed that the responsibility of the lifecycle has moved.  With the approach of the lifecycle of the chaincode being controlled independently of the peer what is being offered with this feature to the developer? 

Is the intent to give them just an API that they can use to write their own logic to start up and configure?  ( so they could for example link this in with say Spring Boot, or Quarkus?)

Today the chaincode is in effect a CLI application that is started but by the peer. Is there a requirement here for somebody to 'just want to start' the code. 

Is there a high-level overview of how this feature is used please? I'll track back.

Reply:
@mbwhite 
```
Is the intent to give them just an API that they can use to write their own logic to start up and configure? ( so they could for example link this in with say Spring Boot, or Quarkus?)

Today the chaincode is in effect a CLI application that is started but by the peer. Is there a requirement here for somebody to 'just want to start' the code.
```
Correct. We just give them the API and they deploy and manage as they like (the driving use case was to be able to manage in cloud say as a k8s service - https://jira.hyperledger.org/browse/FAB-13582).

This doc should give an overview of chaincode-as-serverhttps://hyperledger-fabric.readthedocs.io/en/latest/cc_service.html (there's an update PR https://github.com/hyperledger/fabric/pull/419). The "chaincode" there could be in any language.

Reply:
now, i create GrpcServerSetting in this data class we store all properties.
and in example Application i read ENV, and fill this class